### PR TITLE
[POC] C++ optimizer redesign (storing options and states in OptimizerBase)

### PR DESCRIPTION
**Pull request by yf225**
I tried out the following designs for C++/Python optimizer parity, and found the pros and cons of each approach:

**Approach 1**: Storing `state` and `param_group` as a dictionary (e.g. `c10::impl::GenericDict`)

This was my original suggestion which mirrors the design of the Python version. Specifically, we have the following data layout:
```
state {
  at::Tensor -> {
    "sum" -> at::Tensor (stored as at::IValue),
    "step" -> int64_t (stored as at::IValue),
  }
}

param_groups [
  group1 {
    "params" -> at::TensorList (stored as at::IValue)
    "options" -> {
      "learning_rate" -> double (stored as at::IValue)
      "lr_decay" -> double (stored as at::IValue)
      "weight_decay" -> double (stored as at::IValue)
      …
    }
  },
  group2 {...},
  …
]
```
Pros:
1. Closely mirrors the data layout of the Python version.

Cons:
1. Requires conversion from `c10::impl::GenericDict` to the concrete `Options` object every time we need to access the options as an `Options` object (e.g. in the `step` function). And requires conversion from `Options` object to `c10::impl::GenericDict` when we want to store the options into the param groups. These conversions incur a lot of overhead.
2. All values in `state` and `param_groups` are stored as `at::IValue`, which means we need to unpack them every time we need to access them, which incurs a lot of overhead as well.
3. The `at::IValue` container abstraction makes the code not as easy to read and understand.
4. The `c10::impl::GenericDict(c10::StringType::get(), c10::AnyType::get())` data type might not be available in the long term.

-------

**Approach 2**: Represent `state` and `param_group` using specific class types, and store them in `OptimizerBase` class by templatizing `OptimizerBase` class

In light of the shortcomings of Approach 1, this approach is suggested so that we can represent `state` and `param_group` using specific class types:
```
state {
  at::Tensor -> AdagradParamState
    .sum (at::Tensor)
    .step (int64_t)
}

param_groups [
  group1 (AdagradParamGroup)
    .params (std::vector<Tensor>)
    .options (AdagradOptions)
      .learning_rate (double)
      .lr_decay (double)
      .weight_decay (double)
      …
  group2
  …
]
```
and in order to store them in `OptimizerBase` class, we templatize `OptimizerBase` class with:
```cpp
template <typename OptimizerParamState, typename OptimizerParamGroup, typename OptimizerOptions>
class TORCH_API OptimizerBase {
  ...
  OptimizerOptions defaults_;
  std::vector<OptimizerParamGroup> param_groups_;
  ska::flat_hash_map<at::TensorImpl*, OptimizerParamState> state_;
}
```
Pros:
1. No conversion to/from `at::IValue` or `c10::impl::GenericDict` needed, which greatly simplifies the code.
2. Use compile-time checked names to access members of `state` / `param_groups` / `options`, which catches user error much earlier.

Cons:
1. Have to templatize `OptimizerBase` class (and `Optimizer` class), which break backward compatibility for people's code that does `struct MyOptimizer : public Optimizer`.
2. Not able to store optimizers of different classes in the same container (i.e. `std::vector<Optimizer>` cannot be used because `Optimizer` has to be templatized).

-------

**Approach 3 [Implemented in this POC]**: Represent `state` and `param_group` using specific class types, and store them in `OptimizerBase` class using base type pointers

This approach is an improvement over Approach 2. Specifically, it takes advantage of the fact that runtime type information is preserved by pointer to base types (i.e. `std::unique_ptr<BaseType>`'s runtime type can be the derived type, and casting it to the derived type will work). The data layout is as follows:
```
state {
  at::Tensor -> std::unique_ptr<OptimizerParamState> (runtime type: AdagradParamState)
    .sum (at::Tensor)
    .step (int64_t)
}

param_groups [
  group1 (OptimizerParamGroup)
    .params (std::vector<Tensor>)
    .options (std::unique_ptr<OptimizerOptions> (runtime type: AdagradOptions) )
      .learning_rate (double)
      .lr_decay (double)
      .weight_decay (double)
      …
  group2
  …
]
```
Compared to Approach 2 (the pros of Approach 2 also apply to this approach):
Pros:
1. No need to templatize `OptimizerBase` or `Optimizer` class, which has the following benefits:
  a) The user can store various different optimizer types in the same container (e.g. `std::vector<Optimizer>`)
  b) Backward compatibility with the user's existing code
  c) More similar to Python optimizer subclassing syntax (in Python we do `class Adagrad(Optimizer)` to subclass from `Optimizer`).

Cons:
1. We have to static_cast the Options and ParamState references to the concrete types before accessing them in functions such as `step` (which should be a minor annoyance).

cc. @anjali411

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit c3bdae8:
* **1/1** failures introduced in this PR
* **0/1** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=c3bdae83d7f50eeb752966ca5b2997d186fe22c1).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).
